src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -419,16 +490,25 @@ protected FlowNode next(@Nonnull FlowNode current, @Nonnull Collection<FlowNode>
                 // Terminating a parallel scan
                 FlowNode temp = hitParallelStart();
                 if (temp != null) { // Start node for current parallel block now that it is done
+                    nextType = NodeType.PARALLEL_START;
                     return temp;
                 }
             } else if (!blackList.contains(p)) {
+                if (p instanceof BlockStartNode && p.getAction(ThreadNameAction.class) != null) {
+                    nextType = NodeType.PARALLEL_BRANCH_START;
+                } else if (ForkScanner.isParallelEnd(p)) {
+                    nextType = NodeType.PARALLEL_END;
+                } else {
+                    nextType = NodeType.NORMAL;
+                }
```

Comment:
dark magic on multiple ifs.
Would it be possible to replace it by a state machine?